chain/network/src/recorder.rs

```diff
@@ -0,0 +1,218 @@
+use crate::types::PeerMessage;
+use actix::Message;
+use near_primitives::{hash::CryptoHash, network::PeerId};
+use serde::ser::SerializeMap;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use tracing::info;
+
+#[derive(Clone, Copy)]
+pub enum Status {
+    Sent,
+    Received,
+}
+#[derive(Default, Serialize, Deserialize, Debug, Clone)]
+struct CountSize {
+    count: usize,
+    bytes: usize,
+}
+
+impl CountSize {
+    fn update(&mut self, bytes: usize) {
+        self.count += 1;
+        self.bytes += bytes;
+    }
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone)]
+struct SentReceived {
+    sent: CountSize,
+    received: CountSize,
+}
+
+impl SentReceived {
+    fn get(&mut self, status: Status) -> &mut CountSize {
+        match status {
+            Status::Sent => &mut self.sent,
+            Status::Received => &mut self.received,
+        }
+    }
+}
+
+// TODO(MarX): Use more compact debug. Right now it will print all known hashes.
+#[derive(Default, Debug, Deserialize, Clone)]
+struct HashAggregator {
+    total: usize,
+    all: HashSet<CryptoHash>,
+}
+
+impl HashAggregator {
+    fn add(&mut self, hash: CryptoHash) {
+        self.total += 1;
+        self.all.insert(hash);
+    }
+
+    fn different(&self) -> usize {
+        self.all.len()
+    }
+}
+
+impl Serialize for HashAggregator {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut dic = serializer.serialize_map(Some(2))?;
+        dic.serialize_entry("total", &self.total)?;
+        dic.serialize_entry("different", &self.different())?;
+        dic.end()
+    }
+}
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+struct Latency {
+    received: usize,
+    mean_latency: f64,
+    weighted_latency: f64,
+}
+
+impl Latency {
+    fn add(&mut self, latency: f64) {
+        if self.received == 0 {
+            self.weighted_latency = latency;
+        } else {
+            self.weighted_latency = 0.8f64 * self.weighted_latency + 0.2f64 * latency;
```

Comment:
what's this hardcoded magic/

Reply:
It is convex combination of received latencies with exponential decay for old latencies, it is a trick to soft the curve. For sake of clarity I'll move this hardcoded number to a constant.